src/System.Management.Automation/engine/Subsystem/CommandPrediction/ICommandPredictor.cs

```diff
@@ -40,22 +40,35 @@ public interface ICommandPredictor : ISubsystem
         /// A command line was accepted to execute.
         /// The predictor can start processing early as needed with the latest history.
         /// </summary>
+        /// <param name="clientId">Represents the client that initiates the call.</param>
         /// <param name="history">History command lines provided as references for prediction.</param>
-        void StartEarlyProcessing(IReadOnlyList<string> history);
+        void StartEarlyProcessing(string clientId, IReadOnlyList<string> history);
 
         /// <summary>
-        /// The suggestion given by the predictor was accepted.
+        /// Get the predictive suggestions. It indicates the start of a suggestion rendering session.
         /// </summary>
-        /// <param name="acceptedSuggestion">The accepted suggestion text.</param>
-        void OnSuggestionAccepted(string acceptedSuggestion);
+        /// <param name="clientId">Represents the client that initiates the call.</param>
+        /// <param name="context">The <see cref="PredictionContext"/> object to be used for prediction.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the prediction.</param>
+        /// <returns>An instance of <see cref="SuggestionPackage"/>.</returns>
+        SuggestionPackage GetSuggestion(string clientId, PredictionContext context, CancellationToken cancellationToken);
 
         /// <summary>
-        /// Get the predictive suggestions.
+        /// One or more suggestions provided by the predictor were displayed to the user.
         /// </summary>
-        /// <param name="context">The <see cref="PredictionContext"/> object to be used for prediction.</param>
-        /// <param name="cancellationToken">The cancellation token to cancel the prediction.</param>
-        /// <returns>A list of predictive suggestions.</returns>
-        List<PredictiveSuggestion>? GetSuggestion(PredictionContext context, CancellationToken cancellationToken);
+        /// <param name="session">The mini-session where the displayed suggestions came from.</param>
+        /// <param name="countOrIndex">
+        /// When the value is greater than 0, it's the number of displayed suggestions from the list returned in <paramref name="session"/>, starting from the index 0.
+        /// When the value is less than or equal to 0, it means a single suggestion from the list got displayed, and the index is the absolute value.
+        /// </param>
+        void OnSuggestionDisplayed(uint session, int countOrIndex);
+
+        /// <summary>
+        /// The suggestion provided by the predictor was accepted.
+        /// </summary>
+        /// <param name="session">Represents the mini-session where the accepted suggestion came from.</param>
+        /// <param name="acceptedSuggestion">The accepted suggestion text.</param>
+        void OnSuggestionAccepted(uint session, string acceptedSuggestion);
```

Comment:
We could merge both methods if they are always used together.

Reply:
They are not used together. `OnSuggestionDisplayed` will be fired way more than `OnSuggestionAccepted`.

Reply:
My thoughts are if we get 100 suggestions we could weight all 100 (including displayed and accepted ones) and return the result to the prediction service so that it could make more better analyze. So users could prefer a case-sensitive selection or use other sources for selection (local configs, history or other prediction services) to inform the prediction service.

Reply:
> My thoughts are if we get 100 suggestions we could weight all 100 (including displayed and accepted ones) and return the result to the prediction service so that it could make more better analyze.

It's pretty common that we get suggestions from a predictor but user doesn't accept anything, so it's not clear to me how to always include "displayed and accepted ones" in a feedback.

After this change, "OnSuggestionDisplayed" event will be fired very often. I very much want to reduce it ..., so any ideas are very welcome. Can you maybe elaborate your thoughts more?

> So users could prefer a case-sensitive selection or use other sources for selection (local configs, history or other prediction services) to inform the prediction service.

Not sure I understand what you mean, it would be nice if you can clarify more. Basically, we try to avoid sending more user local infor to a predictor than really needed.

Reply:
We've heard more than once that users want to restore a PowerShell session after a restart. For this service it might be useful too.

My thoughts may be vague, but I believe that if a service does not store permanently the (session) state, then this state should be stored locally.

What is a session data? (1) Personal sensitive user data (stored only locally), (2) service data. Service data may be (1) complex data structure (like subset of ML model), (2) simple tips (references on internal service models)

I think that if user data should not be stored on the service, then there might be the following options:
1. No data is stored locally between sessions
2. All session data is stored locally (including model)
2. All sensitive data is stored locally (including service tips). 

Also I think there are many claims about sensing user data to Cloud. The service could be data agnostic and gets encrypted data.

Reply:
This feels like a design proposal for a separate feature "restore a PowerShell session after a restart" :)